### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by cc-a**
# Description

This PR adds developer documentation for the project. As I'm looking to step back from leading the project this is a great time to ask any questions you may have about Dune or the development setup or just about anything so we can record it now.

Fixes: #210 #211 #212 #173

## Type of change

- [X] Documentation (non-breaking change that adds or improves the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [ ] All tests pass (eg. `python -m pytest`)
- [ ] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [ ] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

File: docs/code_overview.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Code Overview
+
+## High-level Architecture
+
+The code base covers several user interfaces covering different aspects of the DUNE run
+control system. There are distinct interfaces but also shared infrastructure and
+approaches.
+
+For this reason the code is organised into three [Django apps]:
```
Comment (dalonsoa) [inline docs/code_overview.md:9]:
```suggestion
For this reason the code is organised into the following [Django apps]:
```
So we can add more in the future minimising the inconsistencies when we - for sure - forget to update this "three" 😄

File: docs/code_overview.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+`python manage.py kafka_consumer`. This makes messages available to the web application
+by storing them in the database. This command is also responsible for pruning messages
+from the database after a configurable retention period.
+
+## Authentication
+
+In the absence of a defined authentication mechanism for use in production, the project
+currently uses the built-in [Django authentication system]. This requires manual
+creation of users via the Django command line interface and stores user data in the
+application database. This is a placeholder that will need to be replaced.
```
Comment (dalonsoa) [inline docs/code_overview.md:76]: Maybe highlight this.
```suggestion
application database. **This is a placeholder that will need to be replaced**.
```

File: docs/code_overview.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+## Permissions
+
+Permissions are used to control the actions that can be performed or information that
+can be viewed by different users. Permissions are defined on the custom user model
+defined in `main/models.py`. A fine-grained approach is taken to permissions whereby
+individual actions are have their own associated permission. Once the authorisation
```
Comment (dalonsoa) [inline docs/code_overview.md:83]:
```suggestion
individual actions have their own associated permission. Once the authorisation
```

File: docs/index.md [hunk 4]
```diff
@@ -1 +1,114 @@
...
+Take the services down with `docker compose down` or by pressing Ctrl+C in the
+corresponding terminal.
+
+## Controller UI
+
+You must use the full Docker Compose profile when working with the Controller UI.
+
+### Getting Started with the Controller UI
+
+Check you have value of the `COMPOSE_PROFILES` variable set to `drunc` (see above). As
```
Comment (dalonsoa) [inline docs/index.md:71]: Maybe clarify that the controller will not work correctly with the `drunc-lite` profile.

File: docs/index.md [hunk 5]
```diff
@@ -1 +1,114 @@
...
+with the Process Manager UI start the Docker Compose services and create a user account
+if not already done. You can then boot a full test session:
+
+```bash
+docker compose exec drunc /boot_test_session.sh
```
Comment (dalonsoa) [inline docs/index.md:76]: I'd add a description here of what this script is actually doing to help developers understand why they are running what they are running.

File: docs/index.md [hunk 6]
```diff
@@ -1 +1,114 @@
...
+```
+
+_Note that the above consumes several Gb of memory._
+
+Once booted you can interact with the root controller via:
+
+```bash
+docker compose exec drunc /entrypoint.sh drunc-controller-shell grpc://localhost:3333
+```
```
Comment (dalonsoa) [inline docs/index.md:85]: This bit here is a bit odd as this serves no purpose when developing the UI - except for debugging purposes, which is what you have been doing. I'd add a separate section about debugging, troubleshooting, etc.

**Review by dalonsoa — APPROVED** [review]
I've made a couple of comments, but looks great to me. 

Certainly part of a different PRs, but I think it would be valuable at some point to describe in detail the Dockerfiles and the docker-compose file. Much (all?) of what this tool does - at least in development - depends on the docker setup, so a good understanding on what's going on there would be useful.

File: docs/index.md (hunk 5 above)
Comment (cc-a) [reply docs/index.md:76]: Done

File: docs/index.md (hunk 6 above)
Comment (cc-a) [reply docs/index.md:85]: True, it's also no longer valid as you need to go via the connectivity service in the new version of drunc. I've removed it.

Comment (cc-a) [reply docs/index.md]: Still a debugging/troubleshooting section is a good idea. I've opened an issue for it - https://github.com/ImperialCollegeLondon/drunc_ui/issues/249

File: docs/index.md (hunk 4 above)
Comment (cc-a) [reply docs/index.md:72]: Done

**Comment by cc-a** [comment]
Sounds good @dalonsoa. I've made #251 to track.

**Review by dalonsoa — APPROVED** [review]
LGTM!

**Review by AdrianDAlessandro — APPROVED** [review]
This looks good to me. I'm not sure what's missing, but what is there works. I've made a comment about removing detail from the readme. Hopefully my suggestion can help get the setup instructions back into there.

**Review by jamesturner246 — APPROVED** [review]
Besides the Mac note @AdrianDAlessandro mentioned, LGTM. :+1: 

I can't think of anything else required to get going offhand. My only questions really were about when and where we should be accessing drunc directly for debugging, but that will be covered in #249, I guess.

File: docs/index.md [hunk 7]
```diff
@@ -1 +1,111 @@
-# Drunc Web User Interface
+# Drunc Web User Interface (drunc-ui)
+
+This repo defines the web interface for various drunc tools. Including:
+
+- [The Process Manager]
+- [The Controllers]
+
+## Running the Project Locally
```
[the remark this answers is not in the document]
Comment (cc-a) [reply docs/index.md]: Done. Found this neat ability to exclude content from a snippet - https://facelessuser.github.io/pymdown-extensions/extensions/snippets/#snippet-lines

File: docs/index.md [hunk 8]
```diff
@@ -1 +1,111 @@
...
+_Note that the above consumes several Gb of memory._
+
+The test session boots a full set of processes and controllers suitable for UI testing
+and development. Once booted you can view the status of the root controller with the UI
+at <http://localhost:8000/controller>.
+
+For details of working with the controller see the [drunc wiki]. You should also be able
+to see the booted processes in the Process Manager UI.
+
+## Not Using Docker
```
[the remark this answers is not in the document]
Comment (cc-a) [reply docs/index.md]: Amended.

File: docs/index.md [hunk 9]
```diff
@@ -1 +1,111 @@
...
+```
+
+_Note that the above consumes several Gb of memory._
+
+The test session boots a full set of processes and controllers suitable for UI testing
+and development. Once booted you can view the status of the root controller with the UI
+at <http://localhost:8000/controller>.
+
+For details of working with the controller see the [drunc wiki]. You should also be able
+to see the booted processes in the Process Manager UI.
```
[the remark this answers is not in the document]
Comment (cc-a) [reply docs/index.md]: I'll review as part of #250

**Pull request merged**